src/sys/unix/selector/epoll.rs

```diff
@@ -87,7 +87,18 @@ impl Selector {
         const MAX_SAFE_TIMEOUT: u128 = libc::c_int::max_value() as u128;
 
         let timeout = timeout
-            .map(|to| cmp::min(to.as_millis(), MAX_SAFE_TIMEOUT) as libc::c_int)
+            // as_millis() truncates, so round up to the nearest millisecond as the documentation
+            // says can happen.  This avoids turning submillisecond timeouts into immediate returns
+            // unless the caller explicitly requests that by specifying a zero timeout.
+            .map(|to| {
+                to.as_millis()
+                    + if to.subsec_nanos() % 1_000_000 == 0 {
+                        0
+                    } else {
+                        1
+                    }
+            })
+            .map(|to| cmp::min(to, MAX_SAFE_TIMEOUT) as libc::c_int)
```

Comment:
We don't need the the remainder, we can just get the `subsec_nanos` directly (just a read of a field). Also let's put into a single `map` closure. Something like the following should would

```suggestion
            .map(|to| cmp::min(
                to.as_millis()
                    // as_millis() truncates, so round up to the nearest millisecond as the documentation
                    // says can happen.  This avoids turning submillisecond timeouts into immediate returns
                    // unless the caller explicitly requests that by specifying a zero timeout.
                    + if to.subsec_nanos() == 0 { 0 } else { 1 }
            , MAX_SAFE_TIMEOUT) as libc::c_int)
```

This will need some manual formatting of the code and the comment.

Reply:
Closures merged and re-rustfmt'ed.

The remainder is not needed for timeouts between 0 and 1 ms, but without it one extra ms would be added to any exact-integer-ms timeout.  The biggest relative error is that a 1ms timeout gets doubled to 2ms.  If the timeouts callers specify are a smooth random variable that's no problem, but humans tend to pick nice round numbers.

The docs do say overrun is possible, and I recognize the desire to avoid a divide instruction, however the syscall a few lines later is probably a lot more expensive, especially on a CPU requiring spectre mitigations.

But it's your API and I have no reason to fight about it, so the remainder can go if you still want it gone.

Reply:
You're right, good catch. Let's do something like the following (will need formatting again):

```rust
            .map(|to| {
                let to_ms = to.as_millis();
                // as_millis() truncates, so round up to the nearest millisecond as the
                // documentation says can happen. This avoids turning submillisecond
                // timeouts into immediate returns unless the caller explicitly requests
                // that by specifying a zero timeout.
                let to_ms = to_ms + if to_ms == 0 && to.subsec_nanos() != 0 { 1 } else { 0 };
                cmp::min(MAX_SAFE_TIMEOUT, to_ms) as libc::c_int 
            })
```

Don't need the remainder and we only do it if the timeout is less than 1 millisecond, but not zero.